# Import packages 

In [1]:
# First, import the relevant modules and packages
import pandas as pd
import numpy as np
from matplotlib import pyplot

In [2]:
# Import Austin_TX_House_Listings_data_v1.csv & us_income_zipcode.csv
df1 = pd.read_csv('../01_raw_data/AustinTXHouseListingsDataV1.csv', low_memory=False)

In [3]:
#Evaluate data
df1.head()

Unnamed: 0.1  Unnamed: 0 address_city          address_neighborhood  \
0             0          15       Austin                           NaN   
1             1          16       Austin                           NaN   
2             2          17       Austin                           NaN   
3             3          18       Austin                           NaN   
4             4          19       Austin  Steiner Ranch-Lakewood Hills   

  address_state    address_streetAddress address_subdivision  address_zipcode  \
0            TX     12801 Wooded Lake Ct                 NaN            78732   
1            TX   904 Lakewood Hills Ter                 NaN            78732   
2            TX        13701 Montview Dr                 NaN            78732   
3            TX   700 Lakewood Hills Ter                 NaN            78732   
4            TX  1008 Lakewood Hills Ter                 NaN            78732   

   bathrooms  bedrooms  ...  latest_salemonth latest_saleyear  \
0        5.0       4.0  ...                 2            2018   
1        5.0       5.0  ...                 8            2020   
2        3.5       NaN  ...                 3            2018   
3        4.0       4.0  ...                 6            2018   
4        4.0       4.0  ...                 8            2019   

  latest_price_source  numOfAccessibilityFeatures  numOfAppliances  numPhotos  \
0      Agent Provided                           0                3         39   
1     Broker Provided                           0                3         58   
2      Agent Provided                           0                0          1   
3      Agent Provided                           0                1         40   
4     Broker Provided                           0                4         40   

   numPriceChanges                                              photo  \
0                1  https://photos.zillowstatic.com/fp/911f9a59acb...   
1                2  https://photos.zillowstatic.com/fp/c6346c4a39d...   
2                1  https://maps.googleapis.com/maps/api/streetvie...   
3                4  https://photos.zillowstatic.com/fp/805b5a6b748...   
4                7  https://photos.zillowstatic.com/fp/a09cfad45a9...   

                                           imagepath  \
0  /content/gdrive/MyDrive/zillow-images/70352485...   
1  /content/gdrive/MyDrive/zillow-images/70352465...   
2  /content/gdrive/MyDrive/zillow-images/83823478...   
3  /content/gdrive/MyDrive/zillow-images/70352478...   
4  /content/gdrive/MyDrive/zillow-images/70352461...   

                                           cleanpath  
0  70352485_911f9a59acb6fe5fd909538a685e2a0c-p_f.jpg  
1  70352465_c6346c4a39d1f87578ee1529329d9296-p_f.jpg  
2  83823478_streetviewlocation13701MontviewDr2CAu...  
3  70352478_805b5a6b748b1d9cffeb93a4c41fedb0-p_f.jpg  
4  70352461_a09cfad45a949f7195c9540224f51e49-p_f.jpg  

[5 rows x 736 columns]

In [4]:
df1.shape

(16482, 736)

It looks like our data has more features (columns) than we need

In [5]:
#Analyze the summary of the data
df1.info

<bound method DataFrame.info of        Unnamed: 0.1  Unnamed: 0 address_city          address_neighborhood  \
0                 0          15       Austin                           NaN   
1                 1          16       Austin                           NaN   
2                 2          17       Austin                           NaN   
3                 3          18       Austin                           NaN   
4                 4          19       Austin  Steiner Ranch-Lakewood Hills   
...             ...         ...          ...                           ...   
16477         16479       22067       Austin                           NaN   
16478         16480       22070       Austin                           NaN   
16479         16481       22072       Austin                           NaN   
16480         16482       22073       Austin                           NaN   
16481         16483       22074       Austin                           NaN   

      address_state    address_

In [6]:
df1.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'address_city', 'address_neighborhood',
       'address_state', 'address_streetAddress', 'address_subdivision',
       'address_zipcode', 'bathrooms', 'bedrooms',
       ...
       'latest_salemonth', 'latest_saleyear', 'latest_price_source',
       'numOfAccessibilityFeatures', 'numOfAppliances', 'numPhotos',
       'numPriceChanges', 'photo', 'imagepath', 'cleanpath'],
      dtype='object', length=736)

In [7]:
# Analyze how the dependent variable, 'latest_price', is correlated with the rest of independent variables
dict(df1.corr()['latest_price'].abs().sort_values(ascending = False))

C:\Users\famma\AppData\Local\Temp\ipykernel_18424\1946049958.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  dict(df1.corr()['latest_price'].abs().sort_values(ascending = False))


{'latest_price': 1.0,
 'taxAssessedValue': 0.8628559938898727,
 'taxAnnualAmount': 0.8156128587000064,
 'bathrooms': 0.551445307264037,
 'resoFactsStats_bathrooms': 0.47586093567052096,
 'livingArea': 0.3919678078297034,
 'resoFactsStats_bathroomsFull': 0.3854093485257164,
 'bedrooms': 0.3236994120035267,
 'resoFactsStats_bedrooms': 0.29769330047900794,
 'schools_2_rating': 0.2879634712029946,
 'schools_1_studentsPerTeacher': 0.2846390962391095,
 'schools_1_rating': 0.24097254565070483,
 'schools_0_rating': 0.23987455502465402,
 'schools_1_size': 0.23598605610932682,
 'resoFactsStats_bathroomsHalf': 0.22839099263704998,
 'stories': 0.21967042532231817,
 'numPhotos': 0.181905441001118,
 'longitude': 0.18157016525279043,
 'hasSpa': 0.17628657766491157,
 'coveredSpaces': 0.17196269771541464,
 'parking': 0.16639348500886164,
 'schools_2_studentsPerTeacher': 0.16305886359458596,
 'garageSpaces': 0.16297483664756376,
 'schools_0_distance': 0.15324071928582075,
 'address_zipcode': 0.152604521

In [8]:
#Create a "corr_r" row to filter out less correlated features (columns)
df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)

C:\Users\famma\AppData\Local\Temp\ipykernel_18424\4094108136.py:2: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)
C:\Users\famma\AppData\Local\Temp\ipykernel_18424\4094108136.py:2: FutureWarning: Behavior when concatenating bool-dtype and numeric-dtype arrays is deprecated; in a future version these will cast to object dtype (instead of coercing bools to numeric values). To retain the old behavior, explicitly cast bool-dtype arrays to numeric dtype.
  df1.loc['corr_r'] = df1.corr()['latest_price'].abs().sort_values(ascending = False)


In [9]:
# Check the new row
df1.tail()

Unnamed: 0.1    Unnamed: 0 address_city address_neighborhood  \
16478    16480.00000  22070.000000       Austin                  NaN   
16479    16481.00000  22072.000000       Austin                  NaN   
16480    16482.00000  22073.000000       Austin                  NaN   
16481    16483.00000  22074.000000       Austin                  NaN   
corr_r       0.12465      0.124756          NaN                  NaN   

       address_state address_streetAddress address_subdivision  \
16478             TX     6318 Clairmont Dr                 NaN   
16479             TX      6104 Abilene Trl                 NaN   
16480             TX      7702 Kincheon Ct                 NaN   
16481             TX         5300 Indio Cv                 NaN   
corr_r           NaN                   NaN                 NaN   

        address_zipcode  bathrooms  bedrooms  ...  latest_salemonth  \
16478      78749.000000   2.000000  4.000000  ...          1.000000   
16479      78749.000000   2.000000  3.000000  ...          1.000000   
16480      78749.000000   2.500000  3.000000  ...          1.000000   
16481      78745.000000        NaN       NaN  ...          1.000000   
corr_r         0.152605   0.551445  0.323699  ...          0.030104   

       latest_saleyear latest_price_source  numOfAccessibilityFeatures  \
16478      2018.000000      Agent Provided                    0.000000   
16479      2018.000000      Agent Provided                    0.000000   
16480      2018.000000      Agent Provided                    0.000000   
16481      2018.000000      Agent Provided                    0.000000   
corr_r        0.059022                 NaN                    0.015614   

        numOfAppliances  numPhotos  numPriceChanges  \
16478          3.000000  33.000000          1.00000   
16479          5.000000  25.000000          5.00000   
16480          1.000000  27.000000          1.00000   
16481          0.000000  11.000000          1.00000   
corr_r         0.048005   0.181905          0.09137   

                                                    photo  \
16478   https://photos.zillowstatic.com/fp/2f26fb52681...   
16479   https://photos.zillowstatic.com/fp/71a9662d303...   
16480   https://photos.zillowstatic.com/fp/f452bc0f4ba...   
16481   https://photos.zillowstatic.com/fp/eb612b3f7c8...   
corr_r                                                NaN   

                                                imagepath  \
16478   /content/gdrive/MyDrive/zillow-images/29487158...   
16479   /content/gdrive/MyDrive/zillow-images/29491564...   
16480   /content/gdrive/MyDrive/zillow-images/29486556...   
16481   /content/gdrive/MyDrive/zillow-images/58314089...   
corr_r                                                NaN   

                                                cleanpath  
16478   29487158_2f26fb52681290dd3e096cb85fc299ec-p_f.jpg  
16479   29491564_71a9662d3031a1834e29f578f5a92b71-p_f.jpg  
16480   29486556_f452bc0f4ba54fe11c89905773f38840-p_f.jpg  
16481   58314089_eb612b3f7c80976715d8a54e389b8e15-p_f.jpg  
corr_r                                                NaN  

[5 rows x 736 columns]

In [10]:
 df1.shape

(16483, 736)

In [11]:
# Exclude columns that have correlation values less than 0.05
df2 = df1.loc[:,df1.loc['corr_r'] >= 0.05]

In [12]:
df2.shape

(16483, 44)

The number of columns are reduced from 736 to 44 by dropping columns that have less than 0.05 correlation value 

In [13]:
df2.head()

Unnamed: 0.1  Unnamed: 0  address_zipcode  bathrooms  bedrooms  dateposted  \
0           0.0        15.0          78732.0        5.0       4.0         NaN   
1           1.0        16.0          78732.0        5.0       5.0         NaN   
2           2.0        17.0          78732.0        3.5       NaN         NaN   
3           3.0        18.0          78732.0        4.0       4.0         NaN   
4           4.0        19.0          78732.0        4.0       4.0         NaN   

    latitude  livingArea  longitude  propertyTaxRate  ...  \
0  30.354322      4060.0 -97.911278             1.98  ...   
1  30.355553      4558.0 -97.912544             1.98  ...   
2  30.382227      3402.0 -97.908333             1.98  ...   
3  30.352081      4749.0 -97.912048             1.98  ...   
4  30.356226      4867.0 -97.911697             1.98  ...   

   schools_1_studentsPerTeacher  schools_1_totalCount  schools_2_distance  \
0                          17.0                   1.0                 6.0   
1                          17.0                   1.0                 6.1   
2                          17.0                   1.0                 5.3   
3                          17.0                   1.0                 6.1   
4                          17.0                   1.0                 6.0   

   schools_2_rating  schools_2_studentsPerTeacher  schools_2_totalCount  \
0               7.0                          17.0                   1.0   
1               7.0                          17.0                   1.0   
2               7.0                          17.0                   1.0   
3               7.0                          17.0                   1.0   
4               7.0                          17.0                   1.0   

   latest_price  latest_saleyear  numPhotos numPriceChanges  
0      715000.0           2018.0       39.0             1.0  
1     1025000.0           2020.0       58.0             2.0  
2       59000.0           2018.0        1.0             1.0  
3      825000.0           2018.0       40.0             4.0  
4      849000.0           2019.0       40.0             7.0  

[5 rows x 44 columns]

In [14]:
# Create a row called 'null_sum_col' to drop columns that have more than 50% missing values
df2.loc['null_sum_col'] = df2.isnull().sum()

C:\Users\famma\AppData\Local\Temp\ipykernel_18424\3052481394.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2.loc['null_sum_col'] = df2.isnull().sum()


In [15]:
df3 = df2.loc[:, df2.loc['null_sum_col'] < 8242]

In [16]:
df3.shape

(16484, 37)

The number of columns further decreased to 37

In [17]:
df4 = df3.dropna()

In [18]:
df4.shape

(12940, 37)

In [19]:
df4['add_zipcode'] = df4.address_zipcode.astype('category')

C:\Users\famma\AppData\Local\Temp\ipykernel_18424\4276552845.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df4['add_zipcode'] = df4.address_zipcode.astype('category')


In [20]:
df4.columns

Index(['Unnamed: 0.1', 'Unnamed: 0', 'address_zipcode', 'bathrooms',
       'bedrooms', 'latitude', 'livingArea', 'longitude', 'propertyTaxRate',
       'resoFactsStats_bathrooms', 'resoFactsStats_bathroomsFull',
       'resoFactsStats_bathroomsHalf', 'resoFactsStats_bedrooms',
       'garageSpaces', 'hasGarage', 'hasSpa', 'hasView', 'parking', 'stories',
       'taxAnnualAmount', 'taxAssessedValue', 'schools_0_distance',
       'schools_0_rating', 'schools_0_size', 'schools_0_studentsPerTeacher',
       'schools_1_rating', 'schools_1_size', 'schools_1_studentsPerTeacher',
       'schools_1_totalCount', 'schools_2_distance', 'schools_2_rating',
       'schools_2_studentsPerTeacher', 'schools_2_totalCount', 'latest_price',
       'latest_saleyear', 'numPhotos', 'numPriceChanges', 'add_zipcode'],
      dtype='object')

In [21]:
# Drop unwanated columns 
#df4 = df4.drop(columns=['Unnamed', 'numPhotos'])
columns_to_drop = ['Unnamed', 'Unnamed','numPhotos']
df4 = df4.drop(columns=columns_to_drop)

KeyError: "['Unnamed', 'Unnamed'] not found in axis"

In [22]:
df4.tail()

Unnamed: 0.1    Unnamed: 0  address_zipcode   bathrooms  \
16477          16479.00000  22067.000000     78749.000000    2.000000   
16478          16480.00000  22070.000000     78749.000000    2.000000   
16479          16481.00000  22072.000000     78749.000000    2.000000   
corr_r             0.12465      0.124756         0.152605    0.551445   
null_sum_col       0.00000      0.000000         0.000000  524.000000   

                bedrooms   latitude   livingArea  longitude  propertyTaxRate  \
16477           3.000000  30.221336  2270.000000 -97.879677         1.980000   
16478           4.000000  30.216406  2645.000000 -97.876045         1.980000   
16479           3.000000  30.213816  1469.000000 -97.873711         1.980000   
corr_r          0.323699   0.070221     0.391968   0.181570         0.067748   
null_sum_col  450.000000   0.000000   106.000000   0.000000        18.000000   

              resoFactsStats_bathrooms  ...  schools_1_totalCount  \
16477                         2.000000  ...              1.000000   
16478                         2.000000  ...              1.000000   
16479                         2.000000  ...              1.000000   
corr_r                        0.475861  ...              0.144867   
null_sum_col                215.000000  ...            389.000000   

              schools_2_distance  schools_2_rating  \
16477                   2.500000          9.000000   
16478                   2.100000          9.000000   
16479                   1.900000          9.000000   
corr_r                  0.060797          0.287963   
null_sum_col          799.000000        799.000000   

              schools_2_studentsPerTeacher schools_2_totalCount latest_price  \
16477                            18.000000             1.000000     425000.0   
16478                            18.000000             1.000000     424999.0   
16479                            18.000000             1.000000     316000.0   
corr_r                            0.163059             0.101158          1.0   
null_sum_col                    799.000000           799.000000          0.0   

             latest_saleyear  numPhotos  numPriceChanges   add_zipcode  
16477            2018.000000  28.000000          1.00000  78749.000000  
16478            2018.000000  33.000000          1.00000  78749.000000  
16479            2018.000000  25.000000          5.00000  78749.000000  
corr_r              0.059022   0.181905          0.09137      0.152605  
null_sum_col        0.000000   0.000000          0.00000      0.000000  

[5 rows x 38 columns]

In [23]:
#df4 = df3.drop(['corr_r', 'null_sum_col']).drop(['null_sum_row'], axis = 'columns')
df4 = df4.drop(['corr_r', 'null_sum_col'])

In [24]:
df4.shape

(12938, 38)

In [25]:
(df4.address_zipcode.min(), df4.address_zipcode.max())

(78617.0, 78759.0)

In [26]:
df4.add_zipcode.dtype

CategoricalDtype(categories=[                0.0, 0.15260452122630824,
                              78617.0,             78652.0,
                              78653.0,             78660.0,
                              78701.0,             78702.0,
                              78703.0,             78704.0,
                              78705.0,             78717.0,
                              78719.0,             78721.0,
                              78722.0,             78723.0,
                              78724.0,             78725.0,
                              78726.0,             78727.0,
                              78728.0,             78729.0,
                              78730.0,             78731.0,
                              78732.0,             78733.0,
                              78734.0,             78735.0,
                              78736.0,             78737.0,
                              78738.0,             78739.0,
                             

In [27]:
# load the us_income data
df1_inc = pd.read_csv('../01_raw_data/us_income_zipcode.csv', low_memory=False)

In [28]:
df1_inc.shape

(364998, 111)

In [31]:
df1_inc.head()

ZIP       Geography Geographic Area Name  Households  \
0  601  860Z200US00601          ZCTA5 00601      5397.0   
1  602  860Z200US00602          ZCTA5 00602     12858.0   
2  603  860Z200US00603          ZCTA5 00603     19295.0   
3  606  860Z200US00606          ZCTA5 00606      1968.0   
4  610  860Z200US00610          ZCTA5 00610      8934.0   

   Households Margin of Error  Households Less Than $10,000  \
0                       264.0                          33.2   
1                       448.0                          27.1   
2                       555.0                          32.1   
3                       171.0                          28.4   
4                       372.0                          20.5   

   Households Less Than $10,000 Margin of Error  \
0                                           4.4   
1                                           2.9   
2                                           2.5   
3                                           5.5   
4                                           2.5   

   Households $10,000 to $14,999  \
0                           15.7   
1                           12.7   
2                           13.4   
3                           13.3   
4                           13.2   

   Households $10,000 to $14,999 Margin of Error  \
0                                            2.9   
1                                            2.1   
2                                            1.6   
3                                            4.4   
4                                            2.5   

   Households $15,000 to $24,999  ...  \
0                           23.9  ...   
1                           20.5  ...   
2                           17.2  ...   
3                           23.3  ...   
4                           23.3  ...   

   Nonfamily Households $150,000 to $199,999  \
0                                        0.0   
1                                        0.0   
2                                        0.6   
3                                        0.0   
4                                        0.0   

   Nonfamily Households $150,000 to $199,999 Margin of Error  \
0                                                2.8           
1                                                1.3           
2                                                0.6           
3                                                7.5           
4                                                1.8           

   Nonfamily Households $200,000 or More  \
0                                    0.0   
1                                    0.0   
2                                    0.2   
3                                    0.0   
4                                    0.0   

   Nonfamily Households $200,000 or More Margin of Error  \
0                                                2.8       
1                                                1.3       
2                                                0.4       
3                                                7.5       
4                                                1.8       

   Nonfamily Households Median Income (Dollars)  \
0                                        9386.0   
1                                       11242.0   
2                                       10639.0   
3                                       15849.0   
4                                       12832.0   

   Nonfamily Households Median Income (Dollars) Margin of Error  \
0                                             1472.0              
1                                             1993.0              
2                                              844.0              
3                                             3067.0              
4                                             2405.0              

   Nonfamily Households Mean Income (Dollars)  \
0                                     13044.0   
1                                     16419.0   
2                           

In [29]:
df2_inc = df1_inc[df1_inc['ZIP'].between(78617,78759)]

In [30]:
df2_inc.shape

(1025, 111)

In [32]:
df1_inc.ZIP.dtype

dtype('int64')

In [ ]:
merged_df = pd.merge(df4, df1_inc, on='Zip Code', how='inner')